### PR TITLE
Add numeric query parser to lucene index

### PR DESCRIPTION
**Pull request by mad**
#1508
-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                return buildNumericQuery(t.field(), t.text(), dataType);
+            } catch (NumberFormatException e) {
+                log.warn("Expected Number type for " + t.field() + ":" + dataType + ", but " + e.getMessage());
+            }
+        }
+
+        return super.newWildcardQuery(t);
+    }
+
+    private Query getNumericRangeQuery(final String field, final Class<?> type, String start, String end, final boolean includeLower,
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: What about other numeric types like `Short` and `Byte`? Will they be supported?

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            }
+            if (!includeUpper) {
+                max = Math.addExact(max, -1);
+            }
+            return LongPoint.newRangeQuery(field, min, max);
+        }
+
+        if (type.equals(Float.class)) {
+            Float min = isMatchAll(start) ? Float.NEGATIVE_INFINITY : Float.parseFloat(start);
+            Float max = isMatchAll(end) ? Float.MAX_VALUE : Float.parseFloat(end);
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: I didn't check the logic but why do you use `NEGATIVE_INFINITY` for `min` and `MAX_VALUE` for `max`?
Should we use `POSITIVE_INFINITY` for `max` or `MIN_VALUE` for `min`?

Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: This is typo

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java (hunk 1 above)
Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: Currently LuceneIndex support two number types Long and Double, see function LuceneIndex.addDocument Number case

I'm simplify numeric type parsing, and now Short and Byte supported too

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    private Query buildNumericQuery(final String field, final String value, Class<?> type) {
+        if (AttributeUtil.isWholeNumber(type)) {
+            if (isMatchAll(value)) {
+                return LongPoint.newRangeQuery(field, Long.MIN_VALUE, Long.MAX_VALUE);
+            }
+            return LongPoint.newExactQuery(field, Long.parseLong(value));
+        } else {
+            if (isMatchAll(value)) {
+                DoublePoint.newRangeQuery(field, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: I am not sure but possibly you missed `return` here

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    @Override
+    protected Query newRangeQuery(final String field, final String start, final String end, final boolean startInclusive,
+                                  final boolean endInclusive) {
+
+        Class<?> dataType = storeRetriever.get(field).getDataType();
+        if (Number.class.isAssignableFrom(dataType)) {
+            try {
+                return getNumericRangeQuery(field, dataType, start, end, startInclusive, endInclusive);
+            } catch (NumberFormatException e) {
+                log.warn("Expected Number type for " + field + ":" + dataType + ", but " + e.getMessage());
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: I am not sure about logging styles but maybe it is better to provide an exception as a separate parameter? I.e. something like:
```
log.warn("Expected Number type for " + field + ":" + dataType + ", but " + " number format was incorrect", e);
```

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java (hunk 3 above)
Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: Yes, you right

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java (hunk 4 above)
Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: Ok, I move this logger to separated method

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            }
+        }
+
+        return super.newTermQuery(t);
+    }
+
+    @Override
+    protected Query newWildcardQuery(final Term t) {
+        if (t.field() == null) {
+            return super.newTermQuery(t);
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: Should this be `super.newWildcardQuery(t)`?

Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: FIxed

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java [hunk 6]
```diff
@@ -697,13 +694,18 @@ private SearchParams convertQuery(Condition<?> condition, final KeyInformation.S
         return params;
     }
 
+    private Function<String, KeyInformation> getField2KeyInformationFunction(final RawQuery query, final KeyInformation.IndexRetriever information) {
```
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java]: @mad Will you remove this method?

**Comment by porunov** [comment]
@mad There was a problem caused by #1506 which was fixed in #1523 . I kindly ask you to rebase your PR against the latest commit in master where the problem with Travis is fixed.

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java [hunk 7]
```diff
@@ -73,6 +69,7 @@
 import java.util.AbstractMap.SimpleEntry;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java]: Remove unused import

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
+// Copyright 2017 JanusGraph Authors
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: Use 2019 year here

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            }
+        }
+
+        return super.newWildcardQuery(t);
+    }
+
+    private Query getNumericRangeQuery(final String field, final Class<?> type, String start, String end, final boolean includeLower,
+                                       final boolean includeUpper) {
+        if (AttributeUtil.isWholeNumber(type)) {
+            Long min = isMatchAll(start) ? Long.MIN_VALUE : Long.parseLong(start);
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: use unwrapped type `long` for min and max

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            Long max = isMatchAll(end) ? Long.MAX_VALUE : Long.parseLong(end);
+            if (!includeLower) {
+                min = Math.addExact(min, 1);
+            }
+            if (!includeUpper) {
+                max = Math.addExact(max, -1);
+            }
+            return LongPoint.newRangeQuery(field, min, max);
+        } else {
+            Double min = isMatchAll(start) ? Double.NEGATIVE_INFINITY : Double.parseDouble(start);
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/NumericTranslationQueryParser.java]: use unwrapped type `double` for min and max

**Comment by porunov** [comment]
Merging this PR by following lazy consensus

**Comment by porunov** [comment]
Thank you @mad for the contribution!

**Pull request merged**